src/hash-join/join_compute_api.h

```diff
@@ -44,166 +49,181 @@ struct join_pair { size_type first, second; };
 {
   if (output_npairs > 0) {
-	size_type* output_data = output.data();
-	auto k = [=] MGPU_DEVICE(size_type index) {
-	  output_data[index] = flip_indices ? joined[index].second : joined[index].first;
-	  output_data[index + output_npairs] = flip_indices ? joined[index].first : joined[index].second;
-	};
-	mgpu::transform(k, output_npairs, context);
+    size_type* output_data = output.data();
+    auto k = [=] MGPU_DEVICE(size_type index) {
+      output_data[index] = flip_indices ? joined[index].second : joined[index].first;
+      output_data[index + output_npairs] = flip_indices ? joined[index].first : joined[index].second;
+    };
+    mgpu::transform(k, output_npairs, context);
   }
 }
 
 
-/// \brief Performs a generic hash based join of columns a and b. Works for both inner and left joins.
-///
-/// \param[in] compute_ctx The CudaComputeContext to shedule this to.
-/// \param[out] out row references into a and b of matching rows
-/// \param[in] a first column to join (left)
-/// \param[in] Number of element in a column (left)
-/// \param[in] b second column to join (right)
-/// \param[in] Number of element in b column (right)
-/// \param[in] additional columns to join (default == NULL)
-/// \param[in] Flag used to reorder the left and right column indices found in the join (default = false)
-template<JoinType join_type,
-  typename input_it,
-  typename input2_it,
-  typename input3_it,
-  typename size_type>
-cudaError_t GenericJoinHash(mgpu::context_t &compute_ctx, mgpu::mem_t<size_type>& joined_output, 
-	const input_it a, const size_type a_count, const input_it b, const size_type b_count,
-	const input2_it a2 = (int*)NULL, const input2_it b2 = (int*)NULL,
-	const input3_it a3 = (int*)NULL, const input3_it b3 = (int*)NULL,
-	bool flip_results = false)
+/* --------------------------------------------------------------------------*/
+/** 
+* @Synopsis  Performs a hash-based join between two sets of gdf_tables.
+* 
+* @Param compute_ctx The Modern GPU context
+* @Param joined_output The output of the join operation
+* @Param left_table The left table to join
+* @Param right_table The right table to join
+* @Param flip_results Flag that indicates whether the left and right tables have been 
+* switched, indicating that the output indices should also be flipped
+* @tparam join_type The type of join to be performed
+* @tparam key_type The data type to be used for the Keys in the hash table
+* @tparam index_type The data type to be used for the output indices
+* 
+* @Returns  cudaSuccess upon successful completion of the join. Otherwise returns
+* the appropriate CUDA error code
+*/
+/* ----------------------------------------------------------------------------*/
+template<JoinType join_type, 
+         typename key_type, 
+         typename index_type,
+         typename size_type>
+cudaError_t compute_hash_join(mgpu::context_t & compute_ctx, 
```

Comment:
This is more  of a comment for the future but I really think if we are making new apis or functions we need to make them stream ready. Even if we provide a default value for the stream where it is = to 0. We are leaving ourselves many places  that we have to go back and change synchronous calls.